### PR TITLE
Adding VPN options for IKE version and IKE split connections

### LINKED ISSUE #3138: StrongSwan with several rightsubnet's - ikev1
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
VR
StrongSwan
~~~

##### CLOUDSTACK VERSION

~~~
ACS 4.11.2.0
~~~

##### CONFIGURATION

Advanced Network
VPC

##### SUMMARY

After upgrading ACS from 4.9.3 (openswan) to 4.11.2 (strongswan), all VPNs with multiple networks have stopped working. Only one of the networks declared in the encryption domain passed traffic.

rightsubnet=192.168.198.0/23,192.168.208.0/23,192.168.170.0/23,192.168.234.0/23,192.168.69.0/24

I changed the configuration manually by creating different Child SAs, one for each network, now all networks work.
https://lists.strongswan.org/pipermail/users/2015-November/008966.html

Example:
#conn for vpn-4.3.2.1
conn vpn-4.3.2.1
 left=1.2.3.4
 leftsubnet=192.168.101.0/24
 right=4.3.2.1
 type=tunnel 
 authby=secret 
 keyexchange=ike
 ike=aes128-sha1-modp1024
 ikelifetime=1h 
 esp=aes128-sha1-modp1024
 lifetime=8h 
 keyingtries=2
 auto=start
 forceencaps=no
 dpddelay=30
 dpdtimeout=120
 dpdaction=restart

conn net-192.168.198.0
 also=vpn-4.3.2.1
 rightsubnet=192.168.198.0/23
 auto=start

conn net-192.168.208.0
 also=vpn-4.3.2.1
 rightsubnet=192.168.208.0/23
 auto=start

conn net-192.168.170.0
 also=vpn-4.3.2.1
 rightsubnet=192.168.170.0/23
 auto=start

conn net-192.168.234.0
 also=vpn-4.3.2.1
 rightsubnet=192.168.234.0/23
 auto=start

conn net-192.168.69.0
 also=vpn-4.3.2.1
 rightsubnet=192.168.69.0/24
 auto=start

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

IKE version allows selecting ike (autoselect), ikev1, or ikev2.
Split connections gives an option of separating the first right subnet from the rest, and kicking out individual statements for each right subnet for better cross-compatibility.

Backported from PR: #4137
update per PR suggestion

Fixes #3138

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [X] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by Pearl1594** [comment]
@blueorangutan package

**Review by rhtyd — APPROVED** [review]
Carry my lgtm from previous PR

**Comment by rhtyd** [comment]
Comment by @kricud from https://github.com/apache/cloudstack/pull/4904#issuecomment-824850119

```
Did testing and found following things.

Problems:

Add VPN Customer Gateway > IKE Version > set:ikev2 > press:ok
GW is created and IKE Version is set to: ike
Workaround: edit VPN Customer Gateway > IKE Version > type:ikev2

Add VPN Customer Gateway > IKE Version > set:ikev1 > press:ok
GW is created and IKE Version is set to: ike
Workaround: edit VPN Customer Gateway > IKE Version > type:ikev1

Add VPN Customer Gateway > Edit VPN Customer Gateway > CIDR list > Command failed due to Internal Server Error
Workaround: Delete VPN Customer Gateway and create again with correct settings.

Suggestions:

Edit help text for Split Connections: Only for IKEv2, whether to split multiple right subnet CIDRs into multiple connection statements.
Source:https://wiki.strongswan.org/projects/strongswan/wiki/FAQ#Multiple-subnets-per-SA
Essence: If you use IKEv1, you need to be a roadwarrior and use the UNITY extension.

Add help text for IKE Version : Connections marked with ike will use IKEv2 when initiating, but accept any protocol version when responding.
Source:https://wiki.strongswan.org/projects/strongswan/wiki/connsection
```

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
Setup:
2 VPCs with 2 n/w tiers each and a VM on each of the n/w tier 

Cases tested:
Case1: 
VPN customer gateway 1 - ike(ikev2) + split connection = false
VPN customer gateway 2 - ike(ikev2) + split connection = false

IPsec tunnel status:
```
ipsec status vpn-10.0.52.84
Routed Connections:
vpn-10.0.52.84{55}:  ROUTED, TUNNEL, reqid 49
vpn-10.0.52.84{55}:   10.1.0.0/16 === 10.2.1.0/24 10.2.2.0/24
Security Associations (1 up, 0 connecting):
vpn-10.0.52.84[28]: ESTABLISHED 13 seconds ago, 10.0.52.83[10.0.52.83]...10.0.52.84[10.0.52.84]
vpn-10.0.52.84{57}:  INSTALLED, TUNNEL, reqid 49, ESP SPIs: c3ffeeb5_i cb883f28_o
vpn-10.0.52.84{57}:   10.1.1.0/24 10.1.2.0/24 === 10.2.1.0/24 10.2.2.0/24


ipsec status vpn-10.0.52.83
Routed Connections:
vpn-10.0.52.83{37}:  ROUTED, TUNNEL, reqid 34
vpn-10.0.52.83{37}:   10.2.0.0/16 === 10.1.1.0/24 10.1.2.0/24
Security Associations (1 up, 0 connecting):
vpn-10.0.52.83[29]: ESTABLISHED 10 seconds ago, 10.0.52.84[10.0.52.84]...10.0.52.83[10.0.52.83]
vpn-10.0.52.83{39}:  INSTALLED, TUNNEL, reqid 36, ESP SPIs: cb883f28_i c3ffeeb5_o
vpn-10.0.52.83{39}:   10.2.1.0/24 10.2.2.0/24 === 10.1.1.0/24 10.1.2.0/24
root@r-24-VM:~# 
```

VPN connection established and all VMs can ping VMs in the other VPCs subnets

Case2:
VPN customer gateway 1 - ike(ikev2) + split connection = true
VPN customer gateway 2 - ike(ikev2) + split connection = true

IPsec tunnel status 
```
root@r-23-VM:~# ipsec status vpn-10.0.52.84
Routed Connections:
vpn-10.0.52.84{21}:  ROUTED, TUNNEL, reqid 20
vpn-10.0.52.84{21}:   10.1.0.0/16 === 10.2.1.0/24
Security Associations (1 up, 0 connecting):
vpn-10.0.52.84[20]: ESTABLISHED 23 seconds ago, 10.0.52.83[10.0.52.83]...10.0.52.84[10.0.52.84]
vpn-10.0.52.84{24}:  INSTALLED, TUNNEL, reqid 20, ESP SPIs: cbf2b114_i c93db1a6_o
vpn-10.0.52.84{24}:   10.1.1.0/24 === 10.2.1.0/24


root@r-24-VM:~# ipsec status vpn-10.0.52.83
Routed Connections:
vpn-10.0.52.83{20}:  ROUTED, TUNNEL, reqid 18
vpn-10.0.52.83{20}:   10.2.0.0/16 === 10.1.1.0/24
Security Associations (1 up, 0 connecting):
vpn-10.0.52.83[20]: ESTABLISHED 17 seconds ago, 10.0.52.84[10.0.52.84]...10.0.52.83[10.0.52.83]
vpn-10.0.52.83{24}:  INSTALLED, TUNNEL, reqid 22, ESP SPIs: c468fcc6_i c5e974ba_o
vpn-10.0.52.83{24}:   10.2.2.0/24 === 10.1.1.0/24
vpn-10.0.52.83{25}:  INSTALLED, TUNNEL, reqid 23, ESP SPIs: c93db1a6_i cbf2b114_o
vpn-10.0.52.83{25}:   10.2.1.0/24 === 10.1.1.0/24
```

VPN connection established between: 
vpc2-n1(10.2.1.0/24) and vpc1-n1 (10.1.1.0/24)
vpc2-n2(10.2.2.0/24) and vpc1-n1 (10.1.1.0/24)

No connection to subnet - 10.1.2.0/24


Case3:
VPN customer gateway 1 - ikev2 + split connection = true ;
VPN customer gateway 2:  ikev2 w/o split connections

IPsec tunnel status:
```
ipsec status vpn-10.0.52.83
Routed Connections:
vpn-10.0.52.83{15}:  ROUTED, TUNNEL, reqid 14
vpn-10.0.52.83{15}:   10.2.0.0/16 === 10.1.1.0/24
Security Associations (1 up, 0 connecting):
vpn-10.0.52.83[15]: ESTABLISHED 4 minutes ago, 10.0.52.84[10.0.52.84]...10.0.52.83[10.0.52.83]
vpn-10.0.52.83{19}:  INSTALLED, TUNNEL, reqid 17, ESP SPIs: c414f83b_i c585fa93_o
vpn-10.0.52.83{19}:   10.2.1.0/24 10.2.2.0/24 === 10.1.1.0/24
root@r-24-VM:~# ipsec status vpn-10.0.52.84
Security Associations (1 up, 0 connecting):


ipsec status vpn-10.0.52.84
Routed Connections:
vpn-10.0.52.84{26}:  ROUTED, TUNNEL, reqid 24
vpn-10.0.52.84{26}:   10.1.0.0/16 === 10.2.1.0/24 10.2.2.0/24
Security Associations (1 up, 0 connecting):
vpn-10.0.52.84[14]: ESTABLISHED 4 minutes ago, 10.0.52.83[10.0.52.83]...10.0.52.84[10.0.52.84]
vpn-10.0.52.84{28}:  INSTALLED, TUNNEL, reqid 24, ESP SPIs: c585fa93_i c414f83b_o
vpn-10.0.52.84{28}:   10.1.1.0/24 === 10.2.1.0/24 10.2.2.0/24
```

Cannot ping to VMs on subnet 10.1.2.0/24 


Test4:
VPN customer gateway 1 : ikev2 + split connection = true 
VPN customer gateway 2 = ikev1

site to site VPN connection gets disconnected
```
2021-04-26 12:49:34,429 DEBUG [c.c.a.t.Request] (RouterStatusMonitor-1:ctx-06786015) (logid:61c6a7aa) Seq 1-7511441228500566349: Sending  { Cmd , MgmtId: 32986808451658, via: 1(trl-176-k-M7-pearl-dsilva-kvm1), Ver: v1, Flags: 100111, [{"com.cloud.agent.api.CheckS2SVpnConnectionsCommand":{"vpnIps":["10.0.52.83"],"accessDetails":{"router.name":"r-24-VM","router.ip":"169.254.189.8"},"wait":"30"}}] }
2021-04-26 12:49:34,653 DEBUG [c.c.a.t.Request] (AgentManager-Handler-16:null) (logid:) Seq 1-7511441228500566349: Processing:  { Ans: , MgmtId: 32986808451658, via: 1, Ver: v1, Flags: 110, [{"com.cloud.agent.api.CheckS2SVpnConnectionsAnswer":{"ipToConnected":{"10.0.52.83":"false"},"ipToDetail":{"10.0.52.83":"IPsec SA not found;Site-to-site VPN have not connected"},"details":"10.0.52.83:11:IPsec SA not found;Site-to-site VPN have not connected&
","result":"true","wait":"0"}}] }
```
image



IPsec tunnel status:
```
ipsec status vpn-10.0.52.83
Routed Connections:
vpn-10.0.52.83{32}:  ROUTED, TUNNEL, reqid 29
vpn-10.0.52.83{32}:   10.2.0.0/16 === 10.1.1.0/24 10.1.2.0/24
Security Associations (0 up, 0 connecting):
  no match
root@r-24-VM:~# 


ipsec status vpn-10.0.52.84
Routed Connections:
vpn-10.0.52.84{50}:  ROUTED, TUNNEL, reqid 44
vpn-10.0.52.84{50}:   10.1.0.0/16 === 10.2.1.0/24
Security Associations (0 up, 0 connecting):
  no match
root@r-23-VM:~# 
```

Get the following in the log file - /var/log/daemon.log:
received NO_PROPOSAL_CHOSEN error notify


There seems to be an issue while testing scenarios  - 3 and 4 - Could you please advice @DaanHoogland @rhtyd

**Comment by kricud** [comment]
@Pearl1594

Case3 and Case4 will not work by design.
VPN settings on both sides must be identical/mach. Only exclusion is DPD it must be enabled or disabled on both sides but in 99% of cases it can have different timers.

Ikev1 and Ikev2 are incompatible.

These scenarios will work:
VPN customer gateway 1 - ike(ikev2) + split connection = false
VPN customer gateway 2 - ike(ikev2) + split connection = false

VPN customer gateway 1 - ike(ikev2) + split connection = true
VPN customer gateway 2 - ike(ikev2) + split connection = true

VPN customer gateway 1 - ike(ikev1) + split connection = false
VPN customer gateway 2 - ike(ikev1) + split connection = false

VPN customer gateway 1 - ike(ike) + split connection = false pasive = true
VPN customer gateway 2 - ike(ikev1) + split connection = false

VPN customer gateway 1 - ike(ike) + split connection = false
VPN customer gateway 2 - ike(ikev2) + split connection = false

VPN customer gateway 1 - ike(ike) + split connection = false
VPN customer gateway 2 - ike(ike) + split connection = false

VPN customer gateway 1 - ike(ike) + split connection = true
VPN customer gateway 2 - ike(ike) + split connection = true

**Comment by Pearl1594** [comment]
Thanks for the confirmation @kricud!

**Comment by rhtyd** [comment]
@kricud if ikev autoselects v2, this case shouldn't work:
```
VPN customer gateway 1 - ike(ike) + split connection = false pasive = true
VPN customer gateway 2 - ike(ikev1) + split connection = false
```
Can you check? Thanks.

**Comment by kricud** [comment]
@rhtyd It will if connection will be initiated from gw2. 
Connections marked with ike will use IKEv2 when initiating, but accept any protocol version when responding.

File: ui/public/locales/pl.json [hunk 1]
```diff
@@ -828,6 +828,7 @@
 "label.icmpcode": "ICMP Code",
 "label.icmptype": "ICMP Type",
 "label.id": "ID",
+"label.ike.version": "IKE Version",
```
Comment (DaanHoogland) [inline ui/public/locales/pl.json:831]: polish version contains mostly english. I'm not sure if we should add to that.

File: systemvm/debian/opt/cloud/bin/ipsectunnel.sh [hunk 2]
```diff
@@ -163,6 +170,20 @@ ipsec_tunnel_add() {
...
+      subnetidx=2
+      for rsubnet in ${rsubnetarr[@]:1}; do
+          sudo echo "" >> $vpnconffile &&
+          sudo echo "conn vpn-$rightpeer-$subnetidx" >> $vpnconffile &&
+          sudo echo "  also=vpn-$rightpeer" >> $vpnconffile &&
+          sudo echo "  auto=route" >> $vpnconffile &&
+          sudo echo "  rightsubnet=$rsubnet" >> $vpnconffile
+          ((++subnetidx))
+      done
+  fi
```
Comment (DaanHoogland) [inline systemvm/debian/opt/cloud/bin/ipsectunnel.sh:185]: separate method, please?

File: systemvm/debian/opt/cloud/bin/ipsectunnel.sh [hunk 3]
```diff
@@ -139,14 +139,21 @@ ipsec_tunnel_add() {
 
   check_and_enable_iptables
 
+  rsubnets="  rightsubnets={$rightnets}"
+  if [ $splitconnections -eq 1 ]
+  then
+      rsubnetarr=(${rightnets})
+      rsubnets="  rightsubnet=${rsubnetarr[0]}"
+  fi
```
Comment (DaanHoogland) [inline systemvm/debian/opt/cloud/bin/ipsectunnel.sh:147]: separate method please?

File: systemvm/debian/opt/cloud/bin/configure.py [hunk 4]
```diff
@@ -595,14 +611,25 @@ def configure_ipsec(self, obj):
...
             time.sleep(1)
 
         # With 'auto=route', connections are established on an attempt to
         # communicate over the S2S VPN. This uses ping to initialize the connection.
-        CsHelper.execute("timeout 5 ping -c 3 %s" % (peerlist.split("/", 1)[0].replace(".0", ".1")))
+        for peer in peerlistarr:
+            octets = peer.split('/', 1)[0].split('.')
+            octets[3] = str((int(octets[3]) + 1))
+            ipinsubnet = '.'.join(octets)
+            CsHelper.execute("timeout 5 ping -c 3 %s" % ipinsubnet)
```
Comment (DaanHoogland) [inline systemvm/debian/opt/cloud/bin/configure.py:632]: factor out in a separate methos, please?

File: systemvm/debian/opt/cloud/bin/configure.py [hunk 5]
```diff
@@ -595,14 +611,25 @@ def configure_ipsec(self, obj):
...
-            if len(result) > 0:
+            done = True
+            for peeridx in range(0, len(peerlistarr)):
+                # Check for the proper connection and subnet
+                conn = rightpeer if not splitconnections else rightpeer if peeridx == 0 else '%s-%d' % (rightpeer, peeridx + 1)
+                result = CsHelper.execute('ipsec status vpn-%s | grep "%s"' % (conn, peerlistarr[peeridx]))
+                # If any of the peers hasn't yet finished, continue the outer loop
+                if len(result) == 0:
+                    done = False
+            if done:
```
Comment (DaanHoogland) [inline systemvm/debian/opt/cloud/bin/configure.py:622]: separate out please?

File: systemvm/debian/opt/cloud/bin/configure.py [hunk 6]
```diff
@@ -582,6 +590,14 @@ def configure_ipsec(self, obj):
...
             file.addeq(" dpdtimeout=120")
             file.addeq(" dpdaction=restart")
+        if splitconnections and peerlistarr.count > 1:
+            logging.debug('Splitting connections for rightsubnets %s' % peerlistarr)
+            for peeridx in range(1, len(peerlistarr)):
+                logging.debug('Adding split connection -%d for subnet %s' % (peeridx + 1, peerlistarr[peeridx]))
+                file.append('')
+                file.search('conn vpn-.*-%d' % (peeridx + 1), "conn vpn-%s-%d" % (rightpeer, peeridx + 1))
+                file.append(' also=vpn-%s' % rightpeer)
+                file.append(' rightsubnet=%s' % peerlistarr[peeridx])
```
Comment (DaanHoogland) [inline systemvm/debian/opt/cloud/bin/configure.py:600]: extra method, please?

File: systemvm/debian/opt/cloud/bin/configure.py [hunk 7]
```diff
@@ -556,10 +556,18 @@ def configure_ipsec(self, obj):
         vpnsecretsfile = "%s/ipsec.vpn-%s.secrets" % (self.VPNCONFDIR, rightpeer)
         ikepolicy = obj['ike_policy'].replace(';', '-')
         esppolicy = obj['esp_policy'].replace(';', '-')
+        splitconnections = obj['split_connections'] if 'split_connections' in obj else False
+        ikeversion = obj['ike_version'] if 'ike_version' in obj and obj['ike_version'].lower() in ('ike', 'ikev1', 'ikev2') else 'ike'
+
+        peerlistarr = peerlist.split(',')
+        if splitconnections:
+            logging.debug('Splitting rightsubnets %s' % peerlistarr)
+            peerlist = peerlistarr[0]
```
Comment (DaanHoogland) [inline systemvm/debian/opt/cloud/bin/configure.py:565]: separate method

File: server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java [hunk 8]
```diff
@@ -494,7 +501,40 @@ public Site2SiteCustomerGateway updateCustomerGateway(UpdateVpnCustomerGatewayCm
...
+                    if (conn.getState() == State.Connected || conn.getState() == State.Error) {
+                        stopVpnConnection(conn.getId());
+                    }
+                    startVpnConnection(conn.getId());
+                } catch (ResourceUnavailableException e) {
+                    // Should never get here, as we are looping on the actual connections, but we must handle it regardless
+                    continue;
+                }
+            }
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/vpn/Site2SiteVpnManagerImpl.java:536]: extra nested methods

**Review by DaanHoogland — COMMENTED** [review]
comment has changed ownership a lot, so my comments should be taken with a grain of salt, but please consider.

**Comment by rhtyd** [comment]
Thanks @kricud, @Pearl1594 can you check the case per @kricud 's comment.

**Comment by Pearl1594** [comment]
@kricud Thanks for the explanation - I had reversed the configuration, hence wasn't able to successfully establish a connection.

Tested all the cases and works as expected.

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by rhtyd** [comment]
Merging this based on smoketests, reviews and manual tests by @Pearl1594 and @kricud - thanks all.

File: api/src/main/java/org/apache/cloudstack/api/command/user/vpn/CreateVpnCustomerGatewayCmd.java [hunk 9]
```diff
@@ -92,6 +92,13 @@
             description = "create site-to-site VPN customer gateway for the project", since = "4.6")
     private Long projectId;
 
+    @Parameter(name = ApiConstants.SPLIT_CONNECTIONS, type = CommandType.BOOLEAN, required = false, description = "For IKEv2, whether to split multiple right subnet cidrs into multiple connection statements.")
+    private Boolean splitConnections;
+
+    @Parameter(name = ApiConstants.IKE_VERSION, type = CommandType.STRING, required = false, description = "Which IKE Version to use, one of ike (autoselect), ikev1, or ikev2. " +
+            "Connections marked with 'ike' will use 'ikev2' when initiating, but accept any protocol version when responding. Defaults to ike")
+    private String ikeVersion;
+
```
Comment (sureshanaparti) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vpn/CreateVpnCustomerGatewayCmd.java:101]: @Pearl1594 add _since_ to both parameters, and validator for ike version.

Comment (sureshanaparti) [reply api/src/main/java/org/apache/cloudstack/api/command/user/vpn/CreateVpnCustomerGatewayCmd.java:101]: Addressed here, https://github.com/apache/cloudstack/pull/4996

**Pull request merged**